packages/workbox-routing/lib/Router.mjs

```diff
@@ -0,0 +1,248 @@
+/*
+ Copyright 2017 Google Inc. All Rights Reserved.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+import core from 'workbox-core';
+import {_private} from 'workbox-core';
+
+import normalizeHandler from './normalizeHandler.mjs';
+
+/**
+ * The Router takes one or more [Routes]{@link Route} and passes each fetch
+ * event through it's routing logic to determine the appropriate way to respond
+ * with a Request.
+ *
+ * If no route matches a given a request, the Router will use a "default"
+ * handler if one is defined.
+ *
+ * Should any of the Route's throw an error, you can define a "catch" handler to
+ * gracefully deal with these issues and respond with a Request.
+ *
+ * If a request matches multiple routes, precedence will be given to the
+ * **earliest** registered route.
+ *
+ * @memberof module:workbox-routing
+ */
+class Router {
+  /**
+   * Constructs a new `Router` instance, without any registered routes.
+   */
+  constructor() {
+    // _routes will contain a mapping of HTTP method name ('GET', etc.) to an
+    // array of all the corresponding Route instances that are registered.
+    this._routes = new Map();
+    this._isListenerRegistered = false;
+  }
+
+  /**
+   * Apply the routing rules to a FetchEvent object to get a Response from an
+   * appropriate handler.
+   *
+   * @param {FetchEvent} event The event passed in to a `fetch` handler.
+   * @return {Promise<Response>|undefined} Returns a promise for a response,
+   * taking the registered routes into account. If there was no matching route
+   * and there's no `defaultHandler`, then returns undefined.
+   */
+  handleRequest(event) {
+    if (process.env.NODE_ENV !== 'production') {
+      core.assert.isInstance(event, FetchEvent, {
+        moduleName: 'workbox-routing',
+        className: 'Router',
+        funcName: 'handleRequest',
+        paramName: 'event',
+      });
+    }
+
+    const url = new URL(event.request.url);
+    if (!url.protocol.startsWith('http')) {
+      if (process.env.NODE_ENV !== 'production') {
+        _private.logger.log(`The URL '${url}' does not start with 'http', so ` +
+          `it can't be handled.`);
+      }
+      return;
+    }
+
+    let {handler, params} = this._findHandlerAndParams(event, url);
```

Comment:
Not blocking. I expected this to be a method more like, `_routeRequest` that could find a Route, call it's handler and return a Response rather than break apart the Route to pass around it's callback like this.

Reply:
The possibility of there being a `defaultHandler` and/or a `catchHandler` complicates what you describe.